text/0119-rust-in-sentry.md

```diff
@@ -0,0 +1,85 @@
+- Start Date: 2023-10-25
+- RFC Type: decision
+- RFC PR: https://github.com/getsentry/rfcs/pull/119
+- RFC Status: draft
+
+# Summary
+
+Make it easier to use Rust code from Sentry/Python.
+
+# Motivation
+
+We want to improve the adoption of Rust within the Sentry/Python codebase, making it easier to do so,
+and removing some hurdles that make current Rust usage very annoying and brittle.
+
+# Background
+
+The current way we use Rust from Sentry/Python has a couple of downsides which we want to improve upon.
+
+These include:
+
+- Maintaining Python Bindings is itself a huge burden and cumbersome to do.
+- Specifically, the introduction of the bindings predates the stable Python ABI.
+- Publishing packages to PyPI means that we have to care about SemVer compatibility and public API.
+- The publishing process itself to PyPI has seen an extremely high failure rate as of late.
+- Getting a Rust dependency into Sentry/Python takes a long chain of publishes/updates and thus takes a long time.
+
+We currently have Rust bindings in the form of the `symbolic` and `sentry-relay`,
+where I believe both python package are built specifically for usage in Sentry, with very limited to no third-party usage.
+In particular, Python `symbolic` contains a subset of the functionality the Rust crate provides,
+and in other cases also exposes functionality that is not part of the Rust crate at all,
+for the sole purpose of having Python bindings for that functionality (example: `proguard`).
+
+# Options Considered
+
+## Maintain crate-specific Python bindings
+
+This approach is listed for completeness, I do _not_ advocate for it.
+With this approach, each Rust crate we are interested in would have its own Python Bindings.
+
+- **pro**: We have smaller more targeted crates/packages to maintain and to publish.
+- **pro/con**: Both pro/con is that each crate/package is in charge of its own workflows.
+- **con**: This would still suffer from the problems around having to maintain a public SemVer API.
+- **con**: It would also rely on publishing to PyPI, and then updating that version within Sentry.
+- **con**: Having multiple smaller Python extension modules will increased the fixed per-module overhead.
+  In particular, each module would have its own bundled copy of the Rust `std`.
+
+## Create a single Sentry-specific bindings package
```

Comment:
i'm in favor of a "other" bindings package for functionality that does not fit into any of the more focussed packages, especially if we are already adding this functionality into packages it doesn't belong into.

I think we would want to keep the arroyo bindings small/focussed because they are potentially useful outside of sentry, therefore I wouldn't want to couple them and their build process to a bunch of native stuff.

my concern with a large bindings crate in general is not even the python wheel size, but having a complicated build process. i think having a mega-package that links rdkafka/breakpad/openssl/... sounds like a recipe for slow CI and hard to debug failures.